### DISCUSSION TITLE
How to receive data 1 character at a time instead of line by line?

### DISCUSSION THREAD

**Discussion by vaelen** [bozimmerman/Zimodem#139] [Q&A] [answered]
When I do something like `ATR0E0S0=1S41=1AT23` (it doesn't matter if I use telnet mode or not) and then telnet to the modem's IP, it connects properly and everything seems fine, but I've noticed that the modem only sends data it has received to the serial port once it receives a line terminator from the socket. 

I think the modem is receiving the data from the socket one byte at a time because I can turn on echo and the echo characters are sent back to the telnet client in real time without waiting for a newline character to be received. This feels like an issue where either the serial port isn't being flushed until a new line is received or the data isn't being queued to the serial port's buffer until a new line is received.

Is this just a configuration option I'm missing? I tried setting ATS40 to a tiny number and using AT&D to clear the delimiter list, but that didn't help either.

**Comment by vaelen** [comment]
@bozimmerman , do you have any ideas about this? I feel like I must be missing something here, because I feel like file transfers wouldn't work if this was always the case.

**Reply by bozimmerman** [reply, marked as answer]
This feels like an issue with the term program.  The firmware definitely does not do that in stream mode. 
Also, ATS40 only applies to packets, which is data sent while still in Command mode.  While in stream mode, ATS40 has no effect.

**Reply by vaelen** [reply]
You're right! I've been using telnet for 25 years, and I had no idea that the telnet protocol had a line mode. To make matters worse, I was connecting to my FreeBSD box via Putty, which ALSO supports telnet line mode. When I disabled line mode in the FreeBSD telnet session to the modem or used netcat instead of telnet, the problem still occurred because Putty was only sending one line at a time to FreeBSD. 

Telnet linemode has a whole RFC, it turns out: https://datatracker.ietf.org/doc/html/rfc1184
